packages/config/src/android/Icon.ts

```diff
@@ -1,24 +1,152 @@
+        iconImage = resizedBackgroundImage.composite(iconImage, 0, 0);
+      }
+      iconImage.write(resolve(dpiFolderPath, IC_LAUNCHER_PNG));
+      iconImage.circle().write(resolve(dpiFolderPath, IC_LAUNCHER_ROUND_PNG));
+    } catch (e) {
+      throw new Error('Encountered an issue resizing app icon: ' + e);
+    }
+  }
+
+  if (!foregroundImage) {
+    // If no foreground image, we shouldn't configure the Android Adaptive Icon
+    return null;
+  }
+
+  // set background color in colors.xml
+  setBackgroundColor(projectRoot, backgroundColor);
+
+  for (let dpi in dpiValues) {
+    const { folderName, scale } = dpiValues[dpi];
+    length = 48 * scale;
+    let dpiFolderPath = resolve(projectRoot, ANDROID_RES_PATH, folderName);
+
+    try {
+      let finalAdpativeIconForeground = await resizeImageAsync(foregroundImage, length);
+      finalAdpativeIconForeground.write(resolve(dpiFolderPath, IC_LAUNCHER_FOREGROUND_PNG));
+
+      if (backgroundImage) {
+        let finalAdpativeIconBackground = await resizeImageAsync(backgroundImage, length);
+        finalAdpativeIconBackground.write(resolve(dpiFolderPath, IC_LAUNCHER_BACKGROUND_PNG));
+      }
+    } catch (e) {
+      throw new Error('Encountered an issue resizing adaptive app icon: ' + e);
+    }
   }
 
-  addWarningAndroid(
-    'icon',
-    'This is the image that your app uses on your home screen, you will need to configure it manually.'
+  // create ic_launcher.xml and ic_launcher_round.xml
+  let icLauncherXmlString = createAdaptiveIconXmlString(
+    backgroundImage ? '' : backgroundColor,
+    backgroundImage
   );
+  await createAdaptiveIconXmlFiles(projectRoot, icLauncherXmlString);
+  return true;
+}
+
+async function resizeImageAsync(imagePath: string, length: number) {
+  let imageBuffer = (await Jimp.read(imagePath)).resize(length, length).quality(100);
+
+  return imageBuffer;
+}
+
+export async function setBackgroundColor(projectDir: string, backgroundColor: string) {
+  let colorsXmlPath = await Colors.getProjectColorsXMLPathAsync(projectDir);
+  if (!colorsXmlPath) {
+    console.warn(
+      'Unable to find a colors.xml file in your android project. Background color is not being set.'
+    );
+    return;
+  }
+  let colorsJson = await Colors.readColorsXMLAsync(colorsXmlPath);
+  let colorItemToAdd = [
+    {
+      _: backgroundColor,
+      $: { name: ICON_BACKGROUND },
+    },
+  ];
+  colorsJson = Colors.setColorItem(colorItemToAdd, colorsJson);
+  Colors.writeColorsXMLAsync(colorsXmlPath, colorsJson);
+}
+
+export const createAdaptiveIconXmlString = (
+  backgroundColor: string | null,
+  backgroundImage: string | null
+) => `<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
+    ${
+      backgroundImage
+        ? `<background android:drawable="@mipmap/ic_launcher_background"/>`
+        : backgroundColor
+        ? `<background android:drawable="@color/iconBackground"/>`
+        : null
```

Comment:
Please add some parenthesis, I think most people are not sure about the operator precedence (`... ?  ... : ...`).

Reply:
@dsokal I'm not sure about this parenthesis idea.
Our linters are removing parenthesis in ternary operators chains and instead they're adding another level of indentation if needed. 🤷 
Moreover I consider below code as readable:
```ts
someConditional
  ? valueIfTruthy
  : anotherConditionalIfFalsy
  ? valueIfTruthy
  : valueIfFalsy
```

Reply:
I strongly disagree that this is readable. We should change the eslint rules.